cloud/vsphere/machineactuator.go

```diff
@@ -556,60 +638,50 @@ func (vc *VsphereClient) GetKubeConfig(master *clusterv1.Machine) (string, error
 	return result, nil
 }
 
-// After master created, move the plugins folder from local to
-// master's .terraform.d/plugins.
-// Currently each init will download the plugin, so for a large number of machines,
-// this will eat up a lot of space.
+// This method exists because during bootstrap some artifacts need to be transferred to the
+// remote master. These include the IP address for the master, terraform state file.
+// In GCE, we set this kind of data as GCE metadata. Unfortunately, vSphere does not have
+// a comparable API.
 func (vc *VsphereClient) SetupRemoteMaster(master *clusterv1.Machine) error {
-	// Because bootstrapping machine and the remote machine can be different platform,
-	// we copy over the master from local to remote, then terraform init for that platform,
-	// and copy the plugins to the remote's .terraform.d directory.
 	machineName := master.ObjectMeta.Name
-	glog.Infof("Setting up the remote master[%s] with terraform config and plugin.", machineName)
+	glog.Infof("Setting up the remote master[%s] with terraform config.", machineName)
 
 	ip, err := vc.GetIP(master)
 	if err != nil {
 		return err
 	}
 
-	glog.Infof("Copying .terraform.d directory to master.")
-	homedir, err := getHomeDir()
+	glog.Infof("Copying the staging directory to master.")
+	machinePath := fmt.Sprintf(MachinePathStageFormat, machineName)
+	_, err = vc.remoteSshCommand(master, fmt.Sprintf("mkdir -p %s", machinePath), "~/.ssh/vsphere_tmp", "ubuntu")
 	if err != nil {
+		glog.Infof("remoteSshCommand failed while creating remote machine stage: %+v", err)
 		return err
 	}
+
 	cmd := exec.Command(
 		"scp", "-i", "~/.ssh/vsphere_tmp",
-		"-o", "StrictHostKeyChecking no",
-		"-o", "UserKnownHostsFile /dev/null",
+		"-o", "StrictHostKeyChecking=no",
+		"-o", "UserKnownHostsFile=/dev/null",
 		"-r",
-		path.Join(homedir, ".terraform.d"),
-		fmt.Sprintf("ubuntu@%s:~/", ip))
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	cmd.Run()
-
-	glog.Infof("Setting up terraform on remote master.")
-	cmd = exec.Command(
-		// TODO: this is taking my private key and username for now.
-		"ssh", "-i", "~/.ssh/vsphere_tmp",
-		"-o", "StrictHostKeyChecking no",
-		"-o", "UserKnownHostsFile /dev/null",
-		fmt.Sprintf("ubuntu@%s", ip),
-		fmt.Sprintf("source ~/.profile; cd ~/.terraform.d/kluster/machines/%s; ~/.terraform.d/bin/terraform init; cp -r ~/.terraform.d/kluster/machines/%s/.terraform/plugins/* ~/.terraform.d/plugins/", machineName, machineName))
+		machinePath,
+		fmt.Sprintf("ubuntu@%s:%s", ip, StageDir))
 	cmd.Stdout = os.Stdout
 	cmd.Stderr = os.Stderr
 	cmd.Run()
 
 	return nil
 }
 
-func (vc *VsphereClient) updateAnnotations(machine *clusterv1.Machine, masterEndpointIp string) error {
+func (vc *VsphereClient) updateAnnotations(machine *clusterv1.Machine, masterEndpointIp, tfState string) error {
```

Comment:
Can you comment on why this is an annotation, as opposed to a field in machine status ProviderStatus? Do you see it moving there eventually, or is there a reason why you don't think it makes sense there?

Reply:
I added a comment.

```
// We are storing these as annotations and not in Machine Status because that's intended for
// "Provider-specific status" that will usually be used to detect updates. Additionally,
// Status requires yet another version API resource which is too heavy to store IP and TF state.
```